control/DomainBlock.js

```diff
@@ -249,16 +256,93 @@ class DomainBlock {
     })
 
     // only add new changed ip addresses, there is no need to remove any old ip addrs
-    for(let addr in set) {
-      if(!existingSet[addr]) {
-        await rclient.saddAsync(key,addr);
+    for (let addr in set) {
+      if (!existingSet[addr]) {
+        await rclient.saddAsync(key, addr);
         let blockSet = "blocked_domain_set";
         if (options.blockSet)
           blockSet = options.blockSet;
         await Block.block(addr, blockSet).catch((err) => undefined);
       }
     }
   }
+  async blockCategory(category, options) {
+    const domains = await this.getCategoryDomains(category);
+    await dnsmasq.addPolicyFilterEntry(domains, options).catch((err) => undefined);
+    sem.emitEvent({
+      type: 'ReloadDNSRule',
+      message: 'DNSMASQ filter rule is updated',
+      toProcess: 'FireMain',
+      suppressEventLogging: true
+    })
+    sem.on('UPDATE_CATEGORY_DYNAMIC_DOMAIN', async (event) => {
```

Comment:
I suggest to add functions in CategoryUpdater.js, which updates category domain list file dynamically. It can be hooked into CategoryUpdater.updateDomain, just as what is done to ip set update.
When dnsmasq-based category block is enabled, simply add a symbolic link in dnsmasq config directory to the corresponding category domain list. This can decouple the category domain list update and category domain block enforcement, the same way as ip set update and IPTables update in IPTables-based category block.